core/src/main/scala/org/typelevel/vault/Key.scala

```diff
@@ -21,17 +21,84 @@
 
 package org.typelevel.vault
 
+import cats.Contravariant
 import cats.Functor
 import cats.effect.kernel.Unique
 import cats.Hash
 import cats.implicits._
+import cats.Invariant
+import cats.data.AndThen
 
 /**
  * A unique value tagged with a specific type to that unique. Since it can only be created as a result of that, it links
  * a Unique identifier to a type known by the compiler.
  */
-final class Key[A] private (private[vault] val unique: Unique.Token) {
+final class Key[A] private (
+  private[vault] val unique: Unique.Token,
+  private[vault] val imapping: InvariantMapping[A]
+) extends InsertKey[A]
+    with LookupKey[A] {
+
+  // Delegates, for convenience.
+  private[vault] type I = imapping.I
+  private[vault] val in = imapping.in
+  private[vault] val out = imapping.out
+
+  // Overloaded ctor to preserve bincompat
+  def this(unique: Unique.Token) =
+    this(unique, InvariantMapping.id[A])
+
+  /**
+   * Create a copy of this key that references the same underlying vault element, transformed from type `B` before
+   * insert, and to `B` after lookup.
+   */
+  def imap[B](f: A => B)(g: B => A): Key[B] =
+    new Key(unique, imapping.imap(f)(g))
+
   override def hashCode(): Int = unique.hashCode()
+
+}
+
+sealed trait InsertKey[-A] { outer =>
+  private[vault] def unique: Unique.Token
+  private[vault] type I
+  private[vault] def in: A => I
+
+  def contramap[B](f: B => A): InsertKey[B] =
```

Comment:
Note that `contramap` and `map` will now be exposed on `Key` itself. Note that it's safe, because you'll end up with an `InsertKey`/`LookupKey` after doing this, but it could be confusing.